klangbecken.liq

```diff
@@ -1,66 +1,115 @@
+#   - KLANGBECKEN_SOCKET_PATH                       #
+#     Path to socket to listen on                   #
+# ================================================= #
+
+# set setting to getenv(var) if set
+def set_if_env(setting, var) =
+  if getenv(var) != "" then
+    set(setting, getenv(var))
+  end
+end
+
+# return value of environment variable if set, otherwise fallback
+def getenv_fallback(var, fallback) =
+  if getenv(var) != "" then
+    getenv(var)
+  else
+    fallback
+  end
+end
 
+# log file
 set("log.file", true)
-set("log.file.path","<syslogdir>/<script>.log")
 set("log.stdout", false)
-set("server.telnet", false)
-set("server.telnet.port", 1234)
-set("log.level", 3)
-
-# set("log.syslog.facility","DAEMON")
-# set("log.syslog.program","liquidsoap")
-# set("log.unix_timestamps",false)
+set_if_env("log.file.path", "KLANGBECKEN_LOG_FILE")
 
+# socket
 set("server.socket", true)
-#set("server.socket.path", "/tmp/klangbecken.sock")
-#set("server.socket.path","<sysrundir>/<script>.sock")
-#set("server.socket.permissions",384)
-#set("server.timeout",30.)
-
-#set("init.daemon.pidfile",true)
-#set("init.daemon.pidfile.path","<sysrundir>/<script>.pid")
-
-#set("harbor.bind_addr","0.0.0.0")
-#set("harbor.max_connections",2)
-#set("harbor.reverse_dns",true)
+set_if_env("server.socket.path", "KLANGBECKEN_SOCKET_PATH")
 
 # Get the Klangbecken data directory
-DATA_DIR = getenv("KLANGBECKEN_DATA")
+DATA_DIR = getenv_fallback("KLANGBECKEN_DATA", "data")
 if not is_directory(DATA_DIR) then
     log(level=2, "ERROR: Cannot find data directory: " ^ DATA_DIR)
     shutdown()
 end
 
 # Get the klangecken command
-CMD_PATH = getenv("KLANGBECKEN_PATH")
-
-# Get alsa device
-ALSA_DEVICE = getenv("KLANGBECKEN_ALSA_DEVICE")
+CMD_PATH = getenv_fallback("KLANGBECKEN_PATH", "./klangbecken.py")
 
+# Get the alsa device
+ALSA_DEVICE = getenv_fallback("KLANGBECKEN_ALSA_DEVICE", "default")
 
 # ================================================= #
 # PLAYLISTS                                         #
 # ================================================= #
 
+# check if track was played recently
+# parses DATA_DIR/index.json and compares with current time
+def check_next_func(r) = 
+  # parse index.json
+  index = file.contents(path.concat(DATA_DIR, "index.json"))
+  index = of_json(default=[("list",[("key","value")])], index)
+  # get id from filename
+  id = basename(request.filename(r))
+  id = string.replace(pattern="\.mp3", fun(_) -> "", id)
+  # get info about track in index
+  meta = list.assoc(default=[], id, index)
+  # get last_play as a unix timestamp
+  last_play = meta["last_play"]
+
+  if string.length(last_play) == 0 then
+    print("track was never played before: " ^ request.filename(r))
+    true
+  else
+    last_play_unix = int_of_string(string.trim(get_process_output("date --date=#{last_play} +%s")))
+    now_unix = int_of_float(gettimeofday())
+    # get current time as a unix timestamp
+    diff = now_unix - last_play_unix
+    if diff < 120 then
```

Comment:
i assume this is not the final value. allowing songs to play in 2min intervals does not sound like what we want :)

Reply:
Yes of course :)
The suggestion from #21 is:

> * Once a day for classics
> * Once in three days for the standard playlist (music)
> * Once an hour for jingles